ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -683,7 +714,147 @@ public void handleReplicaMetadataRequest(NetworkRequest request) throws IOExcept
+    String remoteHostName = replicateBlobRequest.getSourceHostName();
+    // ON_DEMAND_REPLICATION_TODO: Suppose the remote host should use the same type of port as currentNode. Any better way to choice the port?
+    Port remoteHostPort = currentNode.getPortToConnectTo();
+    // ON_DEMAND_REPLICATION_TODO: Add configuration as replicationConfig.replicationConnectionPoolCheckoutTimeoutMs
+    final int connectionPoolCheckoutTimeoutMs = 1000;
+
+    List<PartitionRequestInfo> partitionRequestInfoList = new ArrayList<PartitionRequestInfo>();
+    List<BlobId> blobIds = new ArrayList<BlobId>();
+    blobIds.add(blobId);
+    PartitionRequestInfo partitionInfo = new PartitionRequestInfo(blobId.getPartition(), blobIds);
+    partitionRequestInfoList.add(partitionInfo);
+    // Get the Blob including expired and deleted.
+    GetRequest getRequest = new GetRequest(replicateBlobRequest.getCorrelationId(),
+        replicateBlobRequest.getClientId(), MessageFormatFlags.All, partitionRequestInfoList,
+        GetOption.Include_All);
+    GetResponse getResponse = null;
+    try {
+      ConnectedChannel connectedChannel = connectionPool.checkOutConnection(remoteHostName, remoteHostPort, connectionPoolCheckoutTimeoutMs);;
+      ChannelOutput channelOutput = connectedChannel.sendAndReceive(getRequest);
+      getResponse = GetResponse.readFrom(channelOutput.getInputStream(), clusterMap);
+      if ((getResponse.getError() != ServerErrorCode.No_Error) || (getResponse.getPartitionResponseInfoList() == null)
+          || (getResponse.getPartitionResponseInfoList().size() == 0)) {
+        logger.error("ReplicateBlobRequest failed to get blob {} from the remote node {} {} {}",
+            blobId, remoteHostName, remoteHostPort, getResponse.getError());
+        response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+            ServerErrorCode.Unknown_Error);
+      }
+    } catch (Exception e) {
+      logger.error("ReplicateBlobRequest getBlob {} from the remote node {} {} hit exception ",
+          blobId, remoteHostName, remoteHostPort, e);
+      response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+          ServerErrorCode.Unknown_Error);
+    } finally {
+      if (response != null) {
+        long processingTime = SystemTime.getInstance().milliseconds() - startTime;
+        totalTimeSpent += processingTime;
+        publicAccessLogger.info("{} {} processingTime {}", replicateBlobRequest, response, processingTime);
+        metrics.replicateBlobProcessingTimeInMs.update(processingTime);
+        requestResponseChannel.sendResponse(response, request,
+            new ServerNetworkResponseMetrics(metrics.replicateBlobResponseQueueTimeInMs, metrics.replicateBlobSendTimeInMs,
+                metrics.replicateBlobTotalTimeInMs, null, null, totalTimeSpent));
+        return;
+      }
+    }
```

Comment:
can we merge step 1 and step 2 so we don't have a finally breaking this method.
And in the end, we have to release the input stream in GetResponse since it's Netty ByteBuf.

Reply:
+1

Reply:
Modified. 
It cleaned up the final break but at the same time it adds up the if/else layers. :-)

Reply:
Thanks for the reminder to release the GetResponse.
I'm just curious why nettyByteBufLeakHelper couldn't catch it.